### PR TITLE
Initial release

### LINKED ISSUE #1: The type parser
The core of this SDK is the type extractor function. This issue is about building one.

# Requirements

1. Build a function that accept anything and returns a Json string representing its type.

The resulting types are the following:

- `string`
- `int`
- `float`
- `bool`
- `null`
- `list`
- `object`
- `unknown`

The return format is:

```
{
    propertyType: string;
}
```

2. When parsing lists, extract the types present in the list in an additional field:

```
{
    propertyType: string;
    children: string[]; // e.g. [string, null]
}
```

3. When parsing objects, extract the object fields recursively:

```
{
    propertyType: string;
    children: {
        propertyName: string; // object field's name
        propertyType: string;
        children?: any
    }[];
}
```

4. Build a function that accepts a map of string keys to values of any type and returns an array of objects of the following type:

```
{
    propertyName: string; // map key
    // The other fields are the result of the function build on step 3 applied to the map value 
    propertyType: string;
    children: {
        propertyName: string; // object field's name
        propertyType: string;
        children?: any
    }[];
}
```

Add tests to verify the correct behavior.

# Reference code

We have this logic and tests for it implemented in the Avo Inspector SDKs for other platforms

TypeScript:
The function: https://github.com/avohq/node-avo-inspector/blob/main/src/AvoSchemaParser.ts#L6
Tests: https://github.com/avohq/node-avo-inspector/blob/main/src/__tests__/TestParsing.ts

Java:
The function: https://github.com/avohq/android-avo-inspector/blob/33934530274af6aa7640dfaba54b4d626bbe6e48/avoinspector/src/main/java/app/avo/inspector/AvoSchemaExtractor.java#L172
Tests: https://github.com/avohq/android-avo-inspector/blob/33934530274af6aa7640dfaba54b4d626bbe6e48/avoinspector/src/test/java/app/avo/inspector/MapSimpleTypeSchemaExtractionTests.java

### LINKED ISSUE #2: Avo Inspector entry point
Here we are going to build the public interface that the library users will use. It will invoke the type extractor built in the previous issue and log the results.

# Requirements

- constructor
```
apiKey: string;
env: "prod" | "dev" | "staging";
version: string;
appName: string;
```
should save passed data for later access

- `shouldLog` - a method to enable or disable logs

- `trackSchemaFromEvent(eventName: string, eventProperties: { [propName: string]: any })`: should print the `eventName` parameter and the result of the type extractor function applied to the `eventProperties` parameter if the logs are enabled (see ``shouldLog` method). Should return the result of the type extractor function applied to the `eventProperties` parameter. It should be an async call returning the result of the type extractor function.

Write unit tests for this new interface.
 
# References

https://github.com/avohq/node-avo-inspector/blob/main/src/AvoInspector.ts

### LINKED ISSUE #3: Example app
# Requirements

Create an app that runs locally and can trigger the interface methods with various inputs. Can be  console or UI.

### LINKED ISSUE #4: Send the schemas to the API
The last piece is to send the data we prepared to the API.

# API doc

Endpoint: `api.avo.app/inspector/v1/track`
Method: `POST`

Every time the `trackSchemaFromEvent` method is called we need to send the following to the API:

```
{
  [
    {
      "appName": ...,
      "appVersion": ...,
      "libVersion": ...,
      "libPlatform": "go",
      "apiKey": ...,
      "sessionId": "11332f62-ba2d-11eb-8529-0242ac130003", // UUID
      "messageId": "4cffbdca-ba2e-11eb-8529-0242ac130003", // UUID
      "createdAt": "2021-05-21T09:05:40.735Z",
      "type": "sessionStarted",
    },
    {
      "appName": "...",
      "appVersion": "...",
      "libVersion": "...",
      "libPlatform": "go"
	"sessionId": "11332f62-ba2d-11eb-8529-0242ac130003", //UUID same as in the sessionStarted event
      "messageId": "54cafd77-c6cd-46b0-8459-69b59d44f012", // new UUID
      "createdAt": "2021-05-21T09:05:40.740Z",
      "type": "event",
      // Data we prepared before
      "eventName": "Onboarding Started",
      "eventProperties": [
        {
          "propertyName": "user",
          "propertyType": "object",
          "children": [
            {
              "propertyName": "id",
              "propertyType": "string"
            }
          ]
        },
        {
          "propertyName": "friendNames",
          "propertyType": "list",
           "children": ["string", "null"]
        },
        {
          "propertyName": "isAdmin",
          "propertyType": "bool",
        },
      ]
    },
  ]
}
```

```
### HTTP Responses

- Server returns a `200` response for successful API requests.
- Server returns a `400` response for invalid requests.
Invalid requests include:
    - `Content-type header missing`
    - `Invalid API key
 ```

Write tests for the `trackSchemaFromEvent` method and the payload generated for the API calls.

# Reference

https://github.com/avohq/node-avo-inspector/blob/main/src/AvoNetworkCallsHandler.ts
https://github.com/avohq/node-avo-inspector/blob/main/src/__tests__/TestNetworkCallsHandler.ts
https://github.com/avohq/node-avo-inspector/blob/main/src/__tests__/TestNetworkCallsHandlerPublic.ts

### LINKED ISSUE #5: Implement the sampling rate based on the server response
# Requirements

The success API response is a Json with one field: `samplingRate`.

You need to read it and store. The value before the first successful server response should be 1.0.

Every time before making a call to the API you need to check if a randomly generated number between 0 and 1 is less than current `samplingRate`. If it is - you send the request. If not - you discard the request. See https://github.com/avohq/node-avo-inspector/blob/main/src/AvoNetworkCallsHandler.ts#L68

Write tests for this logic

# Reference

https://github.com/avohq/node-avo-inspector/blob/main/src/AvoNetworkCallsHandler.ts#L113

### LINKED ISSUE #6: Add logs
# Requirements

Go through the node inspector SDK and add similar logging.

The `shouldLog(bool)` interface method should turn the logs on and off.

# Reference

https://github.com/avohq/node-avo-inspector/blob/main/src/AvoInspector.ts#L22

### LINKED ISSUE #7: Release the library
Our clients should be able to easily access the library and include it into their projects with a few lines of code

### LINKED ISSUE #8: Add quick start guide to the Readme
# Reference

https://github.com/avohq/node-avo-inspector/blob/main/README.md

### PR DESCRIPTION
**Pull request by aleks-tpom6oh**
Fixes #1 
Fixes #2 
Fixes #3 
Fixes #4  
Fixes #5  
Fixes #6  
Fixes #7  
Fixes #8

### CODE REVIEW

File: .vscode/settings.json
Comment (hugihlynsson) [inline .vscode/settings.json:1]: Is this helpful for everyone or more of a personalised config? Should we git-ignore it?

File: README.md [hunk 1]
```diff
@@ -1,49 +1,72 @@
...
+Obtain the API key at [Avo.app](https://www.avo.app/welcome)
+
+```go
+import (
+	avoinspector "github.com/avohq/go-avo-inspector"
+)
+
+avoInspector, err := avoinspector.NewAvoInspector(
+    "...", // Your API key obtained in the Avo workspace
+    avoinspector.Dev, // or avoinspector.Stafging, avoinspector.Prod
```
Comment (hugihlynsson) [inline README.md]:
```suggestion
    avoinspector.Dev, // or avoinspector.Staging, avoinspector.Prod
```

File: README.md [hunk 2]
```diff
@@ -1,49 +1,72 @@
...
-  "userId": 1337,
-  "emailAddress": "jane.doe@avo.app",
-  "productId": 45,
-  "revenue": 15.99,
-  "timestamp": 1579263014,
-  "deviceId": "2500-11ec-9621"
-});
 ```
-You can call Avo Inspector SDK with the same parameters as the tracking calls 
+    go get github.com/avohq/go-avo-inspector
```
Comment (hugihlynsson) [inline README.md]:
```suggestion
go get github.com/avohq/go-avo-inspector
```

File: README.md [hunk 3]
```diff
@@ -1,49 +1,72 @@
...
+		"int":  42,
+		"flt":  3.14,
+		"bol":  true,
+		"nul":  nil,
+		"lst":  []interface{}{"foo", "bar", nil, map[string]interface{}{"d": 42}},
+		"obj":  map[string]interface{}{"a": 1, "b": "two", "c": []interface{}{true, 3.14}},
+		"unk":  complex(1, 2),
+		"func": func() {},
+	},
+)
```
Comment (hugihlynsson) [inline README.md]:
```suggestion
result, err := avoInspector.TrackSchemaFromEvent("Test Event", map[string]interface{}{
	"str":  "hello",
	"int":  42,
	"flt":  3.14,
	"bol":  true,
	"nul":  nil,
	"lst":  []interface{}{"foo", "bar", nil, map[string]interface{}{"d": 42}},
	"obj":  map[string]interface{}{"a": 1, "b": "two", "c": []interface{}{true, 3.14}},
	"unk":  complex(1, 2),
	"func": func() {},
})
```

File: README.md [hunk 4]
```diff
@@ -1,49 +1,72 @@
...
+
+Read more in the [Avo documentation](https://www.avo.app/docs/implementation/devs-101#inspecting-events)
+
+### 1.
+
+This method gets actual tracking event parameters, extracts schema automatically and sends it to the Avo Inspector backend.
+It is the easiest way to use the library, just call this method at the same place you call your analytics tools' track methods with the same parameters.
+
+```go
+result, err := avoInspector.TrackSchemaFromEvent("Test Event", map[string]interface{}{
```
Comment (hugihlynsson) [inline README.md:52]: Is the last part a bit strange? Missing some spaces? `map[string]interface{}{`

File: avoInspector_test.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	eventName := "TestEvent"
+	eventProperties := map[string]interface{}{
+		"param1": "value1",
+		"param2": 123,
+	}
+
+	_, err := inspector.TrackSchemaFromEvent(eventName, eventProperties)
+	if fmt.Sprint(err) != "Avo Inspector: schema sending failed: request returned non-200 status code: 400" {
+		t.Errorf("unexpected error: %s", err)
+	}
```
Comment (bjornj12) [inline avoInspector_test.go:106]: nice tests

**Review by bjornj12 — APPROVED** [review]
LGTM, seems like some indentations and spaces might be off as hugi has pointed out. 

I know GO is very opinionated about all that, have your run the go linter and had it sort this out?

File: README.md (hunk 4 above)
Comment (aleks-tpom6oh) [reply README.md:52]: That's how we (me and my AI friend) write go code 😄

**Comment by aleks-tpom6oh** [comment]
> LGTM, seems like some indentations and spaces might be off as hugi has pointed out.

I've ran vscode format on all files and it's fine according to the formatter, so I'm pretty happy with that. Here it is using tabs instead of spaces, that's why GitHub diff looks so weird, it looks way better in the IDE 😄

File: .vscode/settings.json
Comment (aleks-tpom6oh) [reply .vscode/settings.json:1]: You are right, going to exclude it

**Pull request merged**